types/src/user_string.rs

```diff
@@ -0,0 +1,195 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! A "user string" is a string literal in Move code. User strings can be present either within
+//! Move modules or scripts, or as arguments to transactions.
+//!
+//! The name is inspired by the "user string" table described in ECMA-335 [1].
+//!
+//! Within this code base, user strings are represented as a different type to prevent them from
+//! mixing with other sorts of strings. For example, it is not possible to use one as an
+//! identifier for name resolution.
+//!
+//! [1] https://www.ecma-international.org/publications/standards/Ecma-335.htm
+
+use canonical_serialization::{
+    CanonicalDeserialize, CanonicalDeserializer, CanonicalSerialize, CanonicalSerializer,
+};
+use failure::prelude::*;
+#[cfg(any(test, feature = "testing"))]
+use proptest_derive::Arbitrary;
+use serde::{Deserialize, Serialize};
+use std::{borrow::Borrow, fmt, ops::Deref, result, string::FromUtf8Error};
+
+/// An owned string literal in a Move transaction.
+///
+/// For more details, see the module level documentation.
+#[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize, Deserialize)]
+#[cfg_attr(any(test, feature = "testing"), derive(Arbitrary))]
+#[cfg_attr(any(test, feature = "testing"), proptest(no_params))]
+pub struct UserString(Box<str>);
```

Comment:
should this be ImmutableString or something similar? This is more of a long term question I suppose.
I know we don't have a String story in Move yet (and not clear if and when) but if this type defines the way to operate on strings in the VM, I find UserString not a very clear name and the idea of that being a literal only is a bit restrictive (though it is for now and as such the comment can stay that way if you feel like).